File: packages/virage-core/src/eval/metrics.ts
/**
 * Pure retrieval evaluation metric functions.
 * No I/O, no side effects — easy to unit-test in isolation.
 */

import type { EvalResult } from "../interfaces/quality.js";

/**
 * Precision@K: fraction of the top-K retrieved items that are relevant.
 */
export function precisionAtK(
  retrieved: string[],
  relevant: Set<string>,
  k: number,
): number {
  if (k <= 0) return 0;
  const topK = retrieved.slice(0, k);
  const hits = topK.filter((id) => relevant.has(id)).length;
  return hits / k;
}

/**
 * Recall@K: fraction of relevant items found in the top-K retrieved items.
 */
export function recallAtK(
  retrieved: string[],
  relevant: Set<string>,
  k: number,
): number {
  if (relevant.size === 0) return 0;
  const topK = retrieved.slice(0, k);
  const hits = topK.filter((id) => relevant.has(id)).length;
  return hits / relevant.size;
}

/**
 * Reciprocal Rank: 1 / rank of the first relevant result (0 if none).
 */
export function reciprocalRank(
  retrieved: string[],
  relevant: Set<string>,
): number {
  for (let i = 0; i < retrieved.length; i++) {
    if (relevant.has(retrieved[i])) return 1 / (i + 1);
  }
  return 0;
}

/**
 * HitRate@K: 1 if at least one relevant result is in the top-K, else 0.
 */
export function hitRateAtK(
  retrieved: string[],
  relevant: Set<string>,
  k: number,
): number {
  return retrieved.slice(0, k).some((id) => relevant.has(id)) ? 1 : 0;
}

interface PerQueryMetrics {
  precisionAt5: number;
  precisionAt10: number;
  recallAt10: number;
  rr: number;
  hitRateAt5: number;
}

/**
 * Average per-query metrics into a single EvalResult.
 */
export function aggregateEvalResults(perQuery: PerQueryMetrics[]): EvalResult {
  if (perQuery.length === 0) {
    return {
      precisionAt5: 0,
      precisionAt10: 0,
      recallAt10: 0,
      mrr: 0,
      hitRateAt5: 0,
      queriesEvaluated: 0,
    };
  }

  const n = perQuery.length;
  const sum = perQuery.reduce(
    (acc, q) => ({
      precisionAt5: acc.precisionAt5 + q.precisionAt5,
      precisionAt10: acc.precisionAt10 + q.precisionAt10,
      recallAt10: acc.recallAt10 + q.recallAt10,
      rr: acc.rr + q.rr,
      hitRateAt5: acc.hitRateAt5 + q.hitRateAt5,
    }),
    { precisionAt5: 0, precisionAt10: 0, recallAt10: 0, rr: 0, hitRateAt5: 0 },
  );

  return {
    precisionAt5: sum.precisionAt5 / n,
    precisionAt10: sum.precisionAt10 / n,
    recallAt10: sum.recallAt10 / n,
    mrr: sum.rr / n,
    hitRateAt5: sum.hitRateAt5 / n,
    queriesEvaluated: n,
  };
}

/**
 * Compute per-query metrics and aggregate them.
 *
 * @param queries - array of `{ retrievedIds, relevantIds }` per query
 */
export function computeEvalResult(
  queries: Array<{ retrievedIds: string[]; relevantIds: Set<string> }>,
): EvalResult {
  const perQuery: PerQueryMetrics[] = queries.map(
    ({ retrievedIds, relevantIds }) => ({
      precisionAt5: precisionAtK(retrievedIds, relevantIds, 5),
      precisionAt10: precisionAtK(retrievedIds, relevantIds, 10),
      recallAt10: recallAtK(retrievedIds, relevantIds, 10),
      rr: reciprocalRank(retrievedIds, relevantIds),
      hitRateAt5: hitRateAtK(retrievedIds, relevantIds, 5),
    }),
  );

  return aggregateEvalResults(perQuery);
}

File: crates/virage-engine/src/eval/metrics.rs
//! Pure retrieval evaluation metric functions. No I/O, no side effects. Ported from
//! `dist/eval/metrics.js` and `dist/eval/ragbench-hf.js`'s `ndcgAtK` — IR-038 Step 5.

use std::collections::HashSet;

/// Precision@K: fraction of the top-K retrieved items that are relevant.
pub fn precision_at_k(retrieved: &[String], relevant: &HashSet<String>, k: usize) -> f64 {
    if k == 0 {
        return 0.0;
    }
    let top_k = &retrieved[..retrieved.len().min(k)];
    let hits = top_k.iter().filter(|id| relevant.contains(*id)).count();
    hits as f64 / k as f64
}

/// Recall@K: fraction of relevant items found in the top-K retrieved items.
pub fn recall_at_k(retrieved: &[String], relevant: &HashSet<String>, k: usize) -> f64 {
    if relevant.is_empty() {
        return 0.0;
    }
    let top_k = &retrieved[..retrieved.len().min(k)];
    let hits = top_k.iter().filter(|id| relevant.contains(*id)).count();
    hits as f64 / relevant.len() as f64
}

/// Reciprocal Rank: 1 / rank of the first relevant result (0 if none).
pub fn reciprocal_rank(retrieved: &[String], relevant: &HashSet<String>) -> f64 {
    for (i, id) in retrieved.iter().enumerate() {
        if relevant.contains(id) {
            return 1.0 / (i + 1) as f64;
        }
    }
    0.0
}

/// HitRate@K: 1 if at least one relevant result is in the top-K, else 0.
pub fn hit_rate_at_k(retrieved: &[String], relevant: &HashSet<String>, k: usize) -> f64 {
    let top_k = &retrieved[..retrieved.len().min(k)];
    if top_k.iter().any(|id| relevant.contains(id)) {
        1.0
    } else {
        0.0
    }
}

/// NDCG@K with binary relevance (relevance = 1 for every relevant doc). Ported from
/// `ragbench-hf.js`'s `ndcgAtK` — RAGBench ground truth carries no graded relevance.
pub fn ndcg_at_k(retrieved: &[String], relevant: &HashSet<String>, k: usize) -> f64 {
    let mut dcg = 0.0;
    for (i, id) in retrieved.iter().take(k).enumerate() {
        if relevant.contains(id) {
            dcg += 1.0 / (i as f64 + 2.0).log2();
        }
    }
    let ideal_count = relevant.len().min(k);
    let mut idcg = 0.0;
    for i in 0..ideal_count {
        idcg += 1.0 / (i as f64 + 2.0).log2();
    }
    if idcg == 0.0 {
        0.0
    } else {
        dcg / idcg
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn set(ids: &[&str]) -> HashSet<String> {
        ids.iter().map(|s| s.to_string()).collect()
    }

    fn vec_of(ids: &[&str]) -> Vec<String> {
        ids.iter().map(|s| s.to_string()).collect()
    }

    #[test]
    fn precision_recall_hit_rate_basic() {
        let retrieved = vec_of(&["a", "b", "c", "d"]);
        let relevant = set(&["b", "d", "z"]);
        assert_eq!(precision_at_k(&retrieved, &relevant, 4), 0.5);
        assert!((recall_at_k(&retrieved, &relevant, 4) - 2.0 / 3.0).abs() < 1e-9);
        assert_eq!(hit_rate_at_k(&retrieved, &relevant, 1), 0.0);
        assert_eq!(hit_rate_at_k(&retrieved, &relevant, 2), 1.0);
    }

    #[test]
    fn reciprocal_rank_first_hit() {
        let retrieved = vec_of(&["a", "b", "c"]);
        let relevant = set(&["c"]);
        assert!((reciprocal_rank(&retrieved, &relevant) - 1.0 / 3.0).abs() < 1e-9);
        assert_eq!(reciprocal_rank(&retrieved, &set(&["z"])), 0.0);
    }

    #[test]
    fn ndcg_perfect_ranking_is_one() {
        let retrieved = vec_of(&["a", "b", "c"]);
        let relevant = set(&["a", "b"]);
        assert!((ndcg_at_k(&retrieved, &relevant, 3) - 1.0).abs() < 1e-9);
    }

    #[test]
    fn ndcg_no_relevant_is_zero() {
        let retrieved = vec_of(&["a", "b"]);
        assert_eq!(ndcg_at_k(&retrieved, &HashSet::new(), 2), 0.0);
    }

    #[test]
    fn empty_relevant_set_is_zero_everywhere() {
        let retrieved = vec_of(&["a"]);
        let relevant = HashSet::new();
        assert_eq!(recall_at_k(&retrieved, &relevant, 1), 0.0);
    }
}
